dex/pair/src/bot_protection.rs

```diff
@@ -0,0 +1,196 @@
+        }
+
+        let bp_config = self.bp_swap_config().get();
+        let current_block = self.blockchain().get_block_nonce();
+        if bp_config.protect_stop_block < current_block {
+            return;
+        }
+
+        let reserve_in = ctx.get_reserve_in();
+        let reserve_out = ctx.get_reserve_out();
+        if reserve_in == &0u64 && reserve_out == &0u64 {
+            return;
+        }
+
+        let caller = ctx.get_caller();
+        let num_swaps = self.num_swaps_by_address(caller).get();
+        require!(
+            num_swaps < bp_config.max_num_actions_per_address,
+            "too many swaps by address"
+        );
+
+        let final_in = ctx.get_final_input_amount();
+        let amount_in_percent = final_in * PERCENT_MAX / reserve_in;
+        require!(
+            amount_in_percent < bp_config.volume_percent,
+            "swap amount in too large"
+        );
+
+        let final_out = ctx.get_final_output_amount();
+        let amount_out_percent = final_out * PERCENT_MAX / reserve_out;
+        require!(
+            amount_out_percent < bp_config.volume_percent,
+            "swap amount out too large"
+        );
+
+        self.num_swaps_by_address(caller).set(&(num_swaps + 1));
+    }
+
+    fn require_can_proceed_remove(&self, ctx: &RemoveLiquidityContext<Self::Api>) {
+        if self.bp_remove_config().is_empty() {
+            return;
+        }
+
+        let bp_config = self.bp_remove_config().get();
+        let current_block = self.blockchain().get_block_nonce();
+        if bp_config.protect_stop_block < current_block {
+            return;
+        }
+
+        let current_liquidity = ctx.get_lp_token_supply();
+        if current_liquidity == &0u64 {
+            return;
+        }
+
+        let caller = ctx.get_caller();
+        let num_removes = self.num_removes_by_address(caller).get();
+        require!(
+            num_removes < bp_config.max_num_actions_per_address,
+            "too many removes by address"
+        );
+
+        let removed_liquidity = &ctx.get_lp_token_payment().amount;
+        let percent_removed = removed_liquidity * PERCENT_MAX / current_liquidity;
+        require!(
+            percent_removed < bp_config.volume_percent,
+            "remove liquidity too large"
+        );
+
+        self.num_removes_by_address(caller).set(&(num_removes + 1));
+    }
+
+    fn require_can_proceed_add(&self, ctx: &AddLiquidityContext<Self::Api>) {
+        if self.bp_add_config().is_empty() {
+            return;
+        }
+
+        let bp_config = self.bp_add_config().get();
+        let current_block = self.blockchain().get_block_nonce();
+        if bp_config.protect_stop_block < current_block {
+            return;
+        }
+
+        let current_liquidity = ctx.get_lp_token_supply();
+        if current_liquidity == &0u64 {
+            return;
+        }
+
+        let caller = ctx.get_caller();
+        let num_adds = self.num_adds_by_address(caller).get();
```

Comment:
nice addition - but we need to clean the addresses after :). somehow.

Reply:
done